[MM-43042] Skeleton loader and toggle widgets based on scope

BenCookie95 opened the pull request:
#### Summary
This contains all the skeleton loader components we will need and I added them to top channels and top reactions as examples. I also added a Higher Order Component for the widgets since they all share a lot of functionality. Currently the HOC toggles the Card titles and also wraps the child components in a card.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-43106
https://mattermost.atlassian.net/browse/MM-43042

#### Related Pull Requests
<!--
List all PRs related to resolving a ticket. For instance, if you submitted a PR to `mattermost/mattermost-server`, please include it here.
-->
- Has server changes (please link here)
- Has mobile changes (please link here)

#### Screenshots
<img width="1792" alt="Screen Shot 2022-04-06 at 2 24 03 PM" src="https://user-images.githubusercontent.com/14115924/162042875-b463342f-453c-432d-8c73-3b689c22dc3c.png">

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
NONE
```

mkraft commented on components/activity_and_insights/insights/widget_hoc/widget_hoc.tsx:
```diff
@@ -0,0 +1,54 @@
+    return Component;
+}
+
+export default widgetHoc;
```
Is the main purpose of this HOC to lookup the title and subtitle?

BenCookie95 replied:
I also wrap the contents in the card component here rather than adding it to every component. In the next PR I'm also using it for opening opening/closing the modal. The main goal is to stop repeating any logic

BenCookie95 commented on components/activity_and_insights/insights/insights.tsx:
```diff
@@ -46,6 +54,20 @@ const Insights = () => {
+                <TopChannels
+                    size={CardSizes.large}
+                    filterType={filterType}
+                    widgetType={'TOP_CHANNELS'}
```
I'll investigate if I can use this https://reactjs.org/docs/react-component.html#displayname. I agree that I probably don't need both of those props but I may need one

mkraft commented on components/activity_and_insights/insights/card/card.tsx:
```diff
@@ -0,0 +1,58 @@
+import './card.scss';
+
+type Props = {
+    class: string;
```
Yeah, I mean just the purpose of the `class` prop for a component user (developer). It's to customize the styling. Does specifying the class name need to be part of the component API or could we generate a unique style name that can be optionally styled against? Is more or less params better in this case?

BenCookie95 replied:
I can look into generating the classname by trying to utillize: https://reactjs.org/docs/react-component.html#displayname. That could reduce the number of props needed.

Having these props won't dramatically impact performance but the less we have the easier it could be to use. I will also have to consider readability/complexity by doing this but we can assess once it's implemented

mkraft replied:
I don't have a strong opinion on this, just things for you to consider in your choice.

BenCookie95 replied:
I'm going to leave it as is for now but I might revisit it

mkraft reviewed (approved):
Looks good. From my perspective just confirm that https://github.com/mattermost/mattermost-webapp/pull/10147#discussion_r847523676 was the value you wanted and everything was just questions and thoughts for you to do with as you please.

mylonsuren commented on components/activity_and_insights/insights/widget_hoc/widget_hoc.tsx:
```diff
@@ -0,0 +1,54 @@
+    class: string;
+}
+
+function widgetHoc<T>(WrappedComponent: ComponentType<T>) {
+    const Component = (props: T & WidgetHocProps) => {
+        const {formatMessage} = useIntl();
+
+        const title = useCallback(() => {
+            if (props.filterType === InsightsScopes.MY) {
+                return formatMessage(InsightsCardTitles[props.widgetType].myTitle);
+            }
+            return formatMessage(InsightsCardTitles[props.widgetType].teamTitle);
+        }, [props.filterType, props.widgetType]);
+
+        const subTitle = useCallback(() => {
+            if (props.filterType === InsightsScopes.MY) {
+                return formatMessage(InsightsCardTitles[props.widgetType].mySubTitle);
+            }
+            return formatMessage(InsightsCardTitles[props.widgetType].teamSubTitle);
+        }, [props.filterType, props.widgetType]);
+
+        return (
+            <InsightsCard
+                class={props.class}
+                title={title()}
+                subTitle={subTitle()}
+                size={props.size}
+            >
+                <WrappedComponent
+                    {...(props as unknown as T)}
+                />
+            </InsightsCard>
+
+        );
+    };
```
For the title and subtitle, could we use a callback that is passed into the `WrappedComponent` and called with the title and subtitle? 

```
getTitle = (title: string, subTitle: string) => {
     ... use state to update title and subtitle of InsightsCardTitles component
}
...
<WrappedComponent
    getTitle={this.getTitle}, ...otherProps
>
```
Not 100% if this would work, or align with how you're designing these components, but it could help with removing the need for `widgetType`.

BenCookie95 replied:
Good idea but we need to keep widget type for the modal in a later PR. Also, keeping the title/subtitle functions out of the parent component is better due to a separation of concerns and to not bloat the parent component

BenCookie95 commented on components/activity_and_insights/insights/widget_hoc/widget_hoc.tsx:
```diff
@@ -0,0 +1,54 @@
+
+export interface WidgetHocProps {
+    size: CardSize;
+    widgetType: 'TOP_CHANNELS' | 'TOP_REACTIONS';
```
I have these moved into an enum in a later PR

BenCookie95 commented:
@furqanmlk Can you update the split.io feature flag please?

furqanmlk commented:
> @furqanmlk Can you update the split.io feature flag please?

@matthewbirtch Feature flag has been added and Insight page is available to view

matthewbirtch reviewed (changes requested):
Thanks @BenCookie95 a few things:

1. Does this account for all possible loading skeletons? The description on the PR implies that it does, but all I see are the top channels and top reactions skeletons.
2. The skeletons that I had designed had varying lengths/heights for the items. 
<img src="https://user-images.githubusercontent.com/2040554/163279495-929dc834-eb8d-4544-99dd-cc2afed5e4a0.png" width="300">
<img src="https://user-images.githubusercontent.com/2040554/163279509-cd3b95a6-145a-4d62-82e3-de94da4c77a9.png" width="300">

3. As per [the prototype](https://www.figma.com/proto/fKyLulu6MUjQzQa9pBFFrC/Insights-and-Analytics?page-id=548%3A102444&node-id=714%3A139896&viewport=478%2C48%2C0.52&scaling=min-zoom&starting-point-node-id=714%3A139896&show-proto-sidebar=1), the skeleton loaders are supposed to animated (fading in and out). I added comments in the CSS to add the animation.

4. For the circles on the top channels skeleton, they aren't quite showing properly as circles. I see explicit height and width are being set to 16px, but for some reason it's not obeying the height.
<img width="335" alt="image" src="https://user-images.githubusercontent.com/2040554/163283507-9b0ab81a-ff29-4501-a368-a05f9dfa78e5.png">

BenCookie95 commented on components/activity_and_insights/activity_and_insights.scss:
```diff
@@ -1,7 +1,52 @@
+                .top-channel-row {
+                    display: flex;
+                    align-items: center;
+                    margin: 12px 0;
```
The designs have a spacing of 24px so I added 12px of padding above and below each row. Let me know if we want it to be 20 now or leave as 24?

BenCookie95 commented on components/activity_and_insights/insights/skeleton_loader/skeleton_loader.scss:
```diff
@@ -0,0 +1,26 @@
+.skeleton-loader {
+    background-color: rgba(var(--center-channel-color-rgb), 0.08);
```
Thanks for this, I've added it now

BenCookie95 commented:
> Thanks @BenCookie95 a few things:
> 
> 1. Does this account for all possible loading skeletons? The description on the PR implies that it does, but all I see are the top channels and top reactions skeletons.
> 2. The skeletons that I had designed had varying lengths/heights for the items.
> 
> <img alt="" width="300" src="https://user-images.githubusercontent.com/2040554/163279495-929dc834-eb8d-4544-99dd-cc2afed5e4a0.png">
> 
> <img alt="" width="300" src="https://user-images.githubusercontent.com/2040554/163279509-cd3b95a6-145a-4d62-82e3-de94da4c77a9.png">
> 
> 3. As per [the prototype](https://www.figma.com/proto/fKyLulu6MUjQzQa9pBFFrC/Insights-and-Analytics?page-id=548%3A102444&node-id=714%3A139896&viewport=478%2C48%2C0.52&scaling=min-zoom&starting-point-node-id=714%3A139896&show-proto-sidebar=1), the skeleton loaders are supposed to animated (fading in and out). I added comments in the CSS to add the animation.
> 4. For the circles on the top channels skeleton, they aren't quite showing properly as circles. I see explicit height and width are being set to 16px, but for some reason it's not obeying the height.
> 
> <img alt="image" width="335" src="https://user-images.githubusercontent.com/2040554/163283507-9b0ab81a-ff29-4501-a368-a05f9dfa78e5.png">

1. I added the components but they aren't in use right now, I only displayed the top reactions and top channels ones. Sorry for the confusion.
2. Done
3. Done
4. Done

matthewbirtch reviewed (approved):
Nice work! Looks great

furqanmlk reviewed (approved):
LGTM
